### PR TITLE
[A] add sdkbox console

### PR DESCRIPTION
**Pull request by hugohuang1111**

### CODE REVIEW

**Comment by darkdukey** [comment]
@pandamicro @linshun 
This is the PR for integrating sdkbox into cocos package command
Here is the readme document
https://gist.github.com/hugohuang1111/ec2e180ea7cdd5c62cc5

**Comment by natural-law** [comment]
@hugohuang1111 
Does this PR means the command `cocos package` is changed to using `sdk-box`?
As before, 

```
bin/sdkbox
bin/sdkbox.bat
bin/sdkbox.pyc
```

are in the repo: [console-binary](https://github.com/natural-law/console-binary)
And they are downloaded by `download_bin.py`.

If you want to integrate the sdk-box into `cocos package` command. Please:
1. Remove the related files in  [console-binary](https://github.com/natural-law/console-binary)
2. Add sdkbox related files to folder `plugins/plugin_package`

Thank you!

**Comment by linshun** [comment]
The current SDKBox should be divided to two parts: package manager and SDK services.

And the SDK services is base on package manager, and alias to SDKBox cmd line.

For example, if developers input "cocos package" or "cocos package -h", it should be output "cocos package blablabla", but not "sdkbox blablabla".

**Comment by darkdukey** [comment]
Let's focus on functionality first, I don't think how the package manager is architected matters that much as long as user has good experience, that's what it matters

**Comment by linshun** [comment]
Functionality first is ok. 
However, the requirement I mentioned should be finished before our first release. Thanks.

When will it ready in your plan?

**Comment by darkdukey** [comment]
We can change the following
1. help message when user type `cocos package`
2. Remove the files in console-binary, and merge it here instead.

**Comment by darkdukey** [comment]
I think as long as user can use the same command and it gives them expected result, we can always update the internal later.

**Comment by linshun** [comment]
It will be the OFFICIAL package manager as we plan. 
So "cocos package" and "cocos package -h" are just some pieces of it.

I agree that we could update it later, could you please show me the plan and the release point.

**Comment by darkdukey** [comment]
I'll share the plan in mail.

For this PR is there anything else needs to be done, apart from the 2 items, I mentioned?

**Comment by natural-law** [comment]
@darkdukey 
I found the code in `plugin_package`:

```
cmd ='sdkbox ' + ' '.join(argv)
self._run_cmd(cmd)
```

Please use absolute path of `sdkbox` in the cmd. Because if using `sdkbox`, it depends on the environment value `Path`. Maybe it's not found the right file. Thank you!

**Comment by darkdukey** [comment]
Ok @hugohuang1111 will update accordingly.

**Comment by linshun** [comment]
Will sdkbox related files move to folder plugins/plugin_package? or just in bin folder.

**Comment by hugohuang1111** [comment]
https://github.com/natural-law/console-binary/pull/14
pr for https://github.com/natural-law/console-binary

**Comment by natural-law** [comment]
@hugohuang1111 
The PR of console-binary is merged. And I have added a tag v3-console-12.
Please update the version & zip_file_size in `config.json` when you commit changes in cocos2d-console.

**Comment by linshun** [comment]
"1. help message when user type cocos package"
Sounds good.

"For this PR is there anything else needs to be done, apart from the 2 items, I mentioned?"
To be honest, it is better to  rename sdkbox to other name, such as "package", or abstract it as a module. It will be more clear if we treat it as OFFICIAL package manager. 

If you guys insist to use the name of sdkbox, that is fine.
Looking forward to your next plan. Thanks.

**Comment by kaizhao** [comment]
SDKBOX is targeting to solve the sdk integration issue for cocos2d-x. It's THE CURE FOR SDK FATIGUE. 

Its APIs are in the sdkbox namespace. You will need to do sdkbox::init() and put your settings in sdkbox_config.json. You can also use sdkbox.com to remote managing the plugins. 

So what we need to do is make sure that developers understand what SDKBOX is, not renaming it. Otherwise, we will get tons of questions about what is the difference between cocos packager and sdkbox. They are the same, so keep the same name please.

**Comment by kaizhao** [comment]
Plugin APIs are in the sdkbox namespace. You will need to do sdkbox::init() and put your settings in sdkbox_config.json. You can also use sdkbox.com to remote managing the plugins.

So what we need to do is make sure that developers understand what SDKBOX is, not renaming it. Otherwise, we will get tons of questions about what is the difference between cocos package and sdkbox. They are the same, so keep the same name please.

**Comment by kaizhao** [comment]
Plugin APIs are in the sdkbox namespace. You will need to do sdkbox::init() and put your settings in sdkbox_config.json. You can also use sdkbox.com to remote managing the plugins.

Since developers are using SDKBOX anyway, what we need to do is make sure that they are ware of the term. I don't think we should rename it partially in the user flow. Otherwise, we will get tons of questions about what is the difference between cocos package and sdkbox.

**Comment by natural-law** [comment]
@linshun 
The changes of the code is OK for merge. If it's also OK for you, please merge it.
I've no permission to merge PRs. : )

**Comment by ricardoquesada** [comment]
actually I think "cocos package" should be removed from "cocos"... it should be an stand-alone tool.
perhaps a "cpm.py"... "cocos package manager".

"cocos.py" is to manage game projects.
gcpm.py's goal is to manage packages... and it should use sdkbox.py internally.

**Comment by ricardoquesada** [comment]
@kaizhao actually, I think it is bad the sdkbox project to use the term sdkbox in the command line. users will get confused: wait, what is sdkbox? sdkbox is the solution for the sdkbox fatigue or a package manager? simplicity and focus is what sdkbox needs, (and what cocos2d needs).

**Comment by hugohuang1111** [comment]
@natural-law is there something new need modify?

**Comment by ricardoquesada** [comment]
Perhaps there are some misunderstanding. Let me try to clarify them:
- having the sdkbox.py in cocos2d-x is good
- using sdkbox technology as the package manager: nobody is against it
- using "cocos package" just a proxy to sdkbox.py is an issue:
  - it is confusing for cocos2d-x users
  - it is confusing for sdkbox users

Proof. Take a look at the screenshot. I just typed "cocos package -h"... and something completely unexpected appeared. Not user friendly at all for cocos2d-x users.

screen shot 2015-09-16 at 8 53 46 pm

**Comment by darkdukey** [comment]
We can modify the help message

**Comment by hugohuang1111** [comment]
@ricardoquesada 
i remove sdkbox logo, is it better?

**Comment by linshun** [comment]
Please also remove "Import SDKBOX packages into Cocos2d-x projects" when type cocos package -h.

Please review other package commands, and make sure that cocos package is clean. 
It is better to make cocos package simple and clean from SDKBox. Thanks.

**Comment by hugohuang1111** [comment]
@linshun 'sdkbox' string removed

@natural-law can you verify again

File: plugins/plugin_package/plugin_package.py [hunk 1]
```diff
@@ -24,40 +26,19 @@ def brief_description():
...
-        commandObject.run(argv[1:])
+        cmd = self._get_sdkbox_path() + ' --runincocos ' + ' '.join(argv)
+        self._run_cmd(cmd)
+
+    def _run_cmd(self, command, cwd=None):
+        # cocos.CMDRunner.run_cmd(command, False, cwd=cwd)
+        subprocess.call(command, shell=True, cwd=cwd)
+
+    def _get_sdkbox_path(self):
+        path = os.path.dirname(os.path.realpath(__file__))
```
Comment (natural-law) [inline plugins/plugin_package/plugin_package.py]: Please get the path of current file like this:

```
    if getattr(sys, 'frozen', None):
        ret = os.path.realpath(os.path.dirname(sys.executable))
    else:
        ret = os.path.realpath(os.path.dirname(__file__))
```

It's used for maintaining compatibility with the situation: Building python scripts to binary with `pyinstaller`.

**Comment by hugohuang1111** [comment]
@natural-law

**Comment by natural-law** [comment]
@linshun 
It's OK for merge. There isn't any `sdkbox` in the output of `cocos package -h`.

**Comment by mannewalis** [comment]
I suggest that we add an option to sdkbox --brand which you can specify the
name or branding of the tool, i.e. --brand Cocos2d, which will replace all
the strings where SDKBOX is used with Cocos2d. I think this will be easier
to maintain.

Another option is to have separate string localization packs for Cocos2d.
Same idea, change the strings, not the source.

Thoughts?

**Pull request merged**